src/libtls/tls_server.c

```diff
@@ -167,71 +167,120 @@ struct private_tls_server_t {
 	bool curves_received;
 };
 
+/**
+ * Create an array of an intersection of server and peer supported key types
+ */
+static array_t *create_common_key_types(chunk_t hashsig,
+										tls_version_t version_min,
+										tls_version_t version_max)
+{
+	hashtable_t *ht;
+	array_t *key_types;
+	enumerator_t *enumerator;
+	int *k;
+	key_type_t *v, lookup;
+	uint16_t sig_scheme;
+
+	ht = tls_get_supported_key_types(version_min, version_max);
+	if (!ht)
+	{
+		return NULL;
+	}
+	key_types = array_create(sizeof(key_type_t), 1);
+	enumerator = ht->create_enumerator(ht);
+	while (enumerator->enumerate(enumerator, &k, &v))
+	{
+		bio_reader_t *reader;
+
+		reader = bio_reader_create(hashsig);
+		while (reader->remaining(reader) &&
+			   reader->read_uint16(reader, &sig_scheme))
+		{
+			lookup = tls_signature_scheme_to_key_type(sig_scheme);
+			if (*v == lookup)
+			{
+				array_insert(key_types, ARRAY_TAIL, &lookup);
+				break;
+			}
+		}
+		reader->destroy(reader);
+	}
+	enumerator->destroy(enumerator);
+	ht->destroy_function(ht, (void*)free);
+	return key_types;
+}
+
 /**
  * Find a cipher suite and a server key
  */
 static bool select_suite_and_key(private_tls_server_t *this,
 								 tls_cipher_suite_t *suites, int count)
 {
+	array_t *key_types;
+	tls_version_t version_min, version_max;
 	private_key_t *key;
 	key_type_t type;
 
-	key = lib->credmgr->get_private(lib->credmgr, KEY_ANY, this->server,
-									this->server_auth);
+	version_min = this->tls->get_version_min(this->tls);
+	version_max = this->tls->get_version_max(this->tls);
+	key_types = create_common_key_types(this->hashsig, version_min, version_max);
+	if (!key_types || !array_count(key_types))
```

Comment:
The function always returns an array (even was the case before as a hashtable was always returned).